sys/net/application_layer/nanocoap/cache.c

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (C) 2020 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_nanocoap_cache
+ * @{
+ *
+ * @file
+ * @brief       Implementation of common functions for the nanocoap-cache
+ *
+ * @author      Cenk Gündoğan <cenk.guendogan@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "net/nanocoap/cache.h"
+#include "hashes/sha256.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static clist_node_t _cache_list_head = { NULL };
+static clist_node_t _empty_list_head = { NULL };
+
+static nanocoap_cache_entry_t _cache_entries[CONFIG_NANOCOAP_CACHE_ENTRIES];
+
+static nanocoap_cache_replacement_strategy_t _replacement_strategy = NULL;
+
+static int _cache_replacement_lru(void)
+{
+    nanocoap_cache_entry_t *least = NULL;
+    clist_node_t *it = _cache_list_head.next;
+
+    /* no element in the list */
+    if (!it) {
+        return -1;
+    }
+
+    least = (nanocoap_cache_entry_t *) it;
+
+    do {
+        it = it->next;
+        nanocoap_cache_entry_t *nit = (nanocoap_cache_entry_t *) it;
+        if (least->access_time > nit->access_time) {
+            least = nit;
+        }
+    } while (it != _cache_list_head.next);
+
+    return nanocoap_cache_del(least);
+}
+
+void nanocoap_cache_init(void)
+{
+    _cache_list_head.next = NULL;
+    _empty_list_head.next = NULL;
+    memset(_cache_entries, 0, sizeof(_cache_entries));
+    /* construct list of empty entries */
+    for (unsigned i = 0; i < CONFIG_NANOCOAP_CACHE_ENTRIES; i++) {
+        clist_rpush(&_empty_list_head, &_cache_entries[i].node);
+    }
+
+    /* hardcode cache replacement strategy for now */
+    _replacement_strategy = _cache_replacement_lru;
```

Comment:
Why does this need to be dynamic in the first place?

Reply:
Because one might to want to exchange it for better / different caching strategy. But granted, this needs not to be dynamically but could be set at compile time. But from the comment, I don't think any dynamic swap-out is planned here. Maybe it could be set in global context though instead of in the init function.

Reply:
Ping @cgundogan?